### PR TITLE
add signet support

### PR DESCRIPTION
**Pull request by kallewoof**
This PR adds the "signet" network type to the list of networks.

See also:
* https://github.com/rust-bitcoin/constants/pull/8

### CODE REVIEW

**Comment by stevenroose** [comment]
FWIW, https://github.com/rust-bitcoin/rust-bitcoin/pull/255 removed our dependency on rust-bech32-bitcoin but is blocked on a bech32 PR, but that should land soon.

**Comment by stevenroose** [comment]
Getting bitcoin_constants in would make this a lot nicer. It would also solve the problem I have with having to duplicate a lot of code for Elements and Liquid.

**Comment by kallewoof** [comment]
@stevenroose Yeah I noticed a lot of duplicate code all over the place. Wasn't sure what that was about but hopefully it can be reduced. :) I am gonna leave the bech32 PR for now, even if it is not going to relate to this repo specifically in the end.

**Comment by clarkmoody** [comment]
Yeah #255 looks good and should allow us to narrow dependency scope for some projects. I believe we're closing in on [rust-bitcoin/rest-bech32#35](https://github.com/rust-bitcoin/rust-bech32/pull/35).

**Comment by dongcarl** [comment]
@kallewoof I'd be happy to review after rebase :relaxed:

**Comment by kallewoof** [comment]
@dongcarl I didn't realize it was in need of a rebase until you nudged, thanks. I think I've rebased it and it works, but my account is flagged at Travis right now so I don't think that CI will ever finish..

File: src/blockdata/constants.rs
Comment (dongcarl) [inline src/blockdata/constants.rs]: Note to reviewers: you can verify the `time`, `bits`, and `nonce` parameters by checking: 

https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L512

and

https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L270

We need to make sure that, before merging, there's some way of passing in the blockscript and creating the Coinbase signature like so:

https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L505-L513

File: src/consensus/params.rs
Comment (dongcarl) [inline src/consensus/params.rs]: These parameters can be verified here: https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L306

File: src/consensus/params.rs
Comment (dongcarl) [inline src/consensus/params.rs]: The above parameters can be verified here: https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L295-L306

@kallewoof Any particular reason the `BIP{34,65,66}` heights differ from your Core implementation?

File: src/network/constants.rs
Comment (dongcarl) [inline src/network/constants.rs]: Verifiable here: https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L311-L314

File: src/util/bip32.rs
Comment (dongcarl) [inline src/util/bip32.rs]: Changes in this file verifiable here: https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L337-L338

File: src/util/key.rs
Comment (dongcarl) [inline src/util/key.rs]: Verifiable here: https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L336

File: src/util/address.rs
Comment (dongcarl) [inline src/util/address.rs]: Changes in this file verifiable here: https://github.com/kallewoof/bitcoin/blob/2b9db9ac0936cf278ad27c59497d9feec61f6440/src/chainparams.cpp#L334-L340

**Comment by dongcarl** [comment]
Kicking Travis!

File: src/consensus/params.rs
Comment (kallewoof) [reply src/consensus/params.rs]: Mistake; changed here to match (425446c).

**Comment by dr-orlovsky** [comment]
Why is it closed?

**Comment by kallewoof** [comment]
@dr-orlovsky Nobody has looked at it for 5 months. Maybe after signet is merged into core it will have more interest.

**Comment by dr-orlovsky** [comment]
Well, I do need it in rust-bitcoin... Probably I will integrate it into my develop branch directly at this stage.

**Comment by kallewoof** [comment]
I actually didn't realize I can't reopen PRs I close on my own, but in either case, I do intend to make a PR, and I will maintain this branch so feel free to use it! If you need help don't hesitate to ping me btw. I'd love to see what you're working on.

**Comment by dr-orlovsky** [comment]
Thank you! Actually I think it should be possible to re-open PR, at least I've done this myself in the past. It's strange, maybe GitHub changed something since that time.

I'm working on LNP/BP standards and library — layer 2 & 3 solutions on top of Bitcoin and Lightning, including RGB (client-validated rich state management system, like for assets, collectibles, reputation etc), Spectrum (DEX over Lightning), Storm (trustless storage and messaging), Prometheus (trustless computing). You can see more on <https://github.com/lnp-bp/>. The first priority is RGB, which is sort of "Confidential Assets" in Liquid, but with no on-chain data and able to work on LN w/o its modifications and I need Signet to test it, since we are already at pre-release stage (and one of the first asset will be USDT).

**Comment by stevenroose** [comment]
I'm also interested in revisiting this and getting it in. Please reopen PR :)

**Comment by kallewoof** [comment]
@stevenroose Would love to, but the "Reopen" button is grayed out for some reason :)

**Comment by apoelstra** [comment]
When I hover over the "Reopen" button it says the branch was "force-pushed or recreated". I think it won't let us reopen if the branch changed in the time that it was closed. Frustrating.

**Comment by ryanofsky** [comment]
> I think it won't let us reopen if the branch changed in the time that it was closed. Frustrating.

If you reset the branch to 1c0385002bc89ce0d54e000776372c11bc1ec42a, the tip at the time the PR was closed, github will let you reopen

**Comment by kallewoof** [comment]
@ryanofsky Thanks!

File: src/blockdata/constants.rs
Comment (stevenroose) [inline src/blockdata/constants.rs]: Does this mean it's currently incorrect?

Comment (kallewoof) [reply src/blockdata/constants.rs]: No, it means the block hash is not derived automatically, but hard coded. There's a related discussion at https://github.com/bitcoin/bitcoin/pull/16411#issuecomment-577999888 which suggests that we use a hard coded genesis block for all signet instances -- right now, signet derives the genesis block from the challenge (i.e. scriptPubKey of the blocks).

Comment (kallewoof) [reply src/blockdata/constants.rs]: Actually, I take that back. I'm actually unsure about this one. Will have to verify this one.

Comment (kallewoof) [reply src/blockdata/constants.rs]: I've pushed a to-squash commit that properly derives the signet genesis based on the block script. Note that https://github.com/bitcoin/bitcoin/pull/16411#issuecomment-577999888 is still up in the air, but currently keeping things as is.

File: src/blockdata/constants.rs
Comment (stevenroose) [inline src/blockdata/constants.rs]: So this is the default signet signer key?

File: src/blockdata/constants.rs
Comment (stevenroose) [inline src/blockdata/constants.rs]: I think this can be `signet_txdata[0].txid().as_hash().into()` so you can avoid the 2 lines of conversion above.

File: src/blockdata/constants.rs
Comment (stevenroose) [reply src/blockdata/constants.rs]: So currently all signets will have the same genesis hash with the signet0 pubkey in it? Cool. 

ACK if that's what Core is doing.

Comment (kallewoof) [reply src/blockdata/constants.rs]: In practice, it is possible to specify a custom block script which would result in a different genesis hash, but recent discussion indicates we will usually have only one signet at a time, most of the time.

File: src/blockdata/constants.rs
Comment (kallewoof) [reply src/blockdata/constants.rs]: Yes. It may change at some point, but this shouldn't be a common occurrence.

File: src/blockdata/constants.rs
Comment (kallewoof) [reply src/blockdata/constants.rs]: Sounds good

**Comment by kallewoof** [comment]
Due to https://github.com/bitcoin/bips/pull/900, this PR was changed, and the custom genesis block generation stuff was removed.

The new approach requires a dynamic magic number which is not implemented in this PR; instead the hard-coded result is used as is.

**Comment by dr-orlovsky** [comment]
Now will work with electrum (once PR will be accepted): https://github.com/romanz/electrs/pull/239

**Comment by kallewoof** [comment]
@dr-orlovsky Nice! I have https://github.com/Blockstream/electrs/pull/12 which is similar, but I'll close that after comparing notes with your version. Btw, is the romanz repository still active? I assumed not, so I made my PR to the ElementsProject fork.

**Comment by dr-orlovsky** [comment]
Ouch! It is not! I had to work on Blockstream fork instead, it's >200 commits ahead! Seems all the work have to be re-done...

**Comment by elichai** [comment]
I do think it's active. the 200 commits ahead is for liquid support AFAIU @shesek @romanz

**Comment by romanz** [comment]
The main difference between the forks is that [electrs](https://github.com/romanz/electrs/) started as an light-weight indexer for serving personal Electrum wallet using modest hardware (e.g. a Raspberry Pi), and still works well as such :)
[Blockstream fork](https://github.com/Blockstream/electrs/pull/12) is more oriented toward a high-perfomant blockchain explorer - serving as a backend for [Esplora](https://github.com/Blockstream/esplora/) but it requires better hardware (e.g. larger SSD storage and better CPU).

File: src/util/key.rs
Comment (stevenroose) [reply src/util/key.rs:176]: You forgot to also add the parsing logic for this in the `from_wif` method.

**Comment by stevenroose** [comment]
One thing was forgotten, otherwise seems very good. If my comment gets fixed, I think I can ACK.

**Comment by kallewoof** [comment]
@stevenroose good catch! Added Signet case.

**Comment by kallewoof** [comment]
I'm not sure what the crate error is indicating, in Travis:
```
     Running `rustc --crate-name ryu /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/ryu-1.0.5/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=c7e179b45d27fd2b -C extra-filename=-c7e179b45d27fd2b --out-dir /home/travis/build/rust-bitcoin/rust-bitcoin/target/debug/deps -L dependency=/home/travis/build/rust-bitcoin/rust-bitcoin/target/debug/deps --cap-lints allow --cfg integer128`
error: expected identifier, found keyword `crate`
 --> /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/ryu-1.0.5/src/buffer/mod.rs:1:5
  |
1 | use crate::raw;
  |     ^^^^^
error: aborting due to previous error
error: Could not compile `ryu`.
Caused by:
  process didn't exit successfully: `rustc --crate-name ryu /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/ryu-1.0.5/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=c7e179b45d27fd2b -C extra-filename=-c7e179b45d27fd2b --out-dir /home/travis/build/rust-bitcoin/rust-bitcoin/target/debug/deps -L dependency=/home/travis/build/rust-bitcoin/rust-bitcoin/target/debug/deps --cap-lints allow --cfg integer128` (exit code: 101)
warning: build failed, waiting for other jobs to finish...
error: build failed
The command "./contrib/test.sh" exited with 101.
```

File: src/blockdata/constants.rs
Comment (dr-orlovsky) [inline src/blockdata/constants.rs:233]: This test now fails; it works in the following case:

```suggestion
    #[test]
    fn signet_genesis_full_block() {
        let gen = genesis_block(Network::Signet);
        assert_eq!(gen.header.version, 1);
        assert_eq!(gen.header.prev_blockhash, Default::default());
        assert_eq!(format!("{:x}", gen.header.merkle_root),
                  "4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b".to_string());
        assert_eq!(gen.header.time, 1598918400);
        assert_eq!(gen.header.bits, 0x1e0377ae);
        assert_eq!(gen.header.nonce, 52613770);
        assert_eq!(format!("{:x}", gen.header.block_hash()),
                   "00000008819873e925422c1ff0f99f7cc9bbb232af63a077a480a3633bee1ef6".to_string());
    }
```

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]
One test fails + needs rebase with a small conflict in one comment

**Comment by kallewoof** [comment]
Thanks for the nudge. Rebased and fixed genesis nonce/bits/timestamp.

**Review by dr-orlovsky — APPROVED** [review]

**Comment by dr-orlovsky** [comment]
tACK https://github.com/rust-bitcoin/rust-bitcoin/pull/291/commits/a3d9899cb10796dfea4eac9a07e887abcacc0c37

**Comment by dr-orlovsky** [comment]
Not sure however how it match information found in https://github.com/ElementsProject/lightning/pull/2972 about that the new genesis signet nonce is 621297

**Comment by kallewoof** [comment]
@dr-orlovsky Yeah, well noted. The nonce and such unfortunately changed when we nudged the POW limit awhile ago. I haven't done a re-request to the lightning project with an update yet.

**Comment by dr-orlovsky** [comment]
@kallewoof: c-Lightning version PR with fix: https://github.com/ElementsProject/lightning/pull/4068

**Comment by dr-orlovsky** [comment]
Signet got merged into Bitcoin Core master: https://github.com/bitcoin/bitcoin/pull/18267#issuecomment-696357656

Let's merge this PR as well - and unblock downstream Electrs dependency: https://github.com/romanz/electrs/pull/239

**Review by sgeisler — APPROVED** [review]

**Comment by kallewoof** [comment]
Any ideas on how to push this forward? Should I nudge someone, or just sit tight? Who/where if the former?

**Review by elichai — APPROVED** [review]
Verified all the constants against bitcoin core, they all match except `bip16_time` which I couldn't figure out what is the equivalent in bitcoin core

**Comment by kallewoof** [comment]
Thanks, @elichai! I think the bip16_time is only relevant during main net IBD, due to that one offending block, which does not exist in signet, so as long as it's not in the future, it should be fine with any value, I think.

**Pull request merged**
